lumenlearning/lti-middleware
# L3-37 Add middleware/goldilocks validation process and unit tests for LtiOidcUtils

## Description
The middleware now takes the claims from id_token from the LMS and puts them into a new JWT and signs it. This allows for Goldilocks to be able to validate the id_token against the middleware using the middleware's JWKS endpoint, proving that the token did truly come from the middleware.

### Motivation and Context
This change removes the security risk of sending a token from the middleware to goldilocks in which someone could have faked this request. With this change, if someone attempts to fake this request, it will fail the validation.

### Fixes
[L3-37 Implement Middleware side of Goldi/Middleware Validation process](https://lumenlearning.atlassian.net/browse/L3-37)

## How Has This Been Tested?
Note: These steps are specifically for testing locally and in isolation since Goldilocks is not yet available for integration testing.
1. Ensure that the `logging.level.net.unicon=DEBUG` property is present and uncommented in the `application.properties` file before building/running the code. This allows for the JWT that is passed to Goldilocks on the server-side to be output to the logs so that it can be used for further testing.
2. In the LMS, click on the Goldilocks LTI link. Ensure that this still produces the Goldilocks Error UI and that it does not produce any new errors in the logs.
3. In the logs, find the LTI Token value and copy it.
4. Paste the LTI Token into https://jwt.davetonge.co.uk/ along with the Middleware's JWKS endpoint (which is something like https://example.ngrok.io/jwks/jwk). The website should indicate that the JWT is valid, the signature is verified, it is not expired, and the key was found. The payload should match the payload of the JWT from the LMS.

---

## Checklist
